### PR TITLE
RFC: Plugin architecture study.

### PR DESCRIPTION
**Pull request by n1k0**
WiP, this is mostly investigating a possible plugin API for the kinto-admin. Feedback welcome.

### Extending the admin

It's possible to extend the admin by registering plugins. A plugin is an object exposing these properties:

- `routes`: A list of react-router routes to handle
- `reducers`: A map of redux store state nodes and their respective redux reducers
- `sagas`: A list of redux sagas to register

### Sample plugin code

This adds a new view at `test/plugin`, rendering a button which when clicked retrieves an answer asynchronously and renders the result.

```jsx
import React from "react";
import { takeEvery } from "redux-saga";
import { put } from "redux-saga/effects";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";

import renderAdmin from "./app";


const PLUGIN_ACTION_REQUEST = "PLUGIN_ACTION_REQUEST";
const PLUGIN_ACTION_SUCCESS = "PLUGIN_ACTION_SUCCESS";

// Test actions
function testActionRequest() {
  return {type: PLUGIN_ACTION_REQUEST};
}

function testActionSuccess() {
  return {type: PLUGIN_ACTION_SUCCESS, answer: 42};
}

// Container setup
function mapStateToProps(state) {
  return {plugin: state.plugin};
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({testActionRequest}, dispatch);
}

// Plugin view root container component
class TestPlugin extends React.Component {
  render() {
    const {plugin, testActionRequest} = this.props;
    return (
      <div>
        <h1>Hello from plugin!</h1>
        <p>Answer: {plugin.answer}</p>
        <p><button onClick={testActionRequest}>Get answer</button></p>
      </div>
    );
  }
}

const TestPluginContainer = connect(mapStateToProps, mapDispatchToProps)(TestPlugin);

// Test saga
function* testSaga(getState, action) {
  yield new Promise(r => setTimeout(r, 1000));
  yield put(testActionSuccess(42));
}

// Sample fake plugin
const routes = [
  {path: "/test/plugin", components: {content: TestPluginContainer}}
];
const reducers = {
  plugin: function(state = {answer: 0}, action) {
    switch(action.type) {
      case PLUGIN_ACTION_SUCCESS: {
        return {...state, answer: action.answer};
      }
      default: {
        return state;
      }
    }
  }
};
const sagas = [
  [takeEvery, PLUGIN_ACTION_REQUEST, testSaga]
];

// Create plugins array
const plugins = [];
// Register our silly plugin
plugins.push({
  routes,
  reducers,
  sagas,
});

renderAdmin(document.getElementById("app"), plugins);
```

### Enabling plugins

For now and until the API stabilizes, plugins are developed in the kinto-admin repository. To enable one or more plugins, you need to set the `KINTO_ADMIN_PLUGINS` environment variable to list those you want to activate when running the development server:

```
$ KINTO_ADMIN_PLUGINS="example,signoff" npm start
```

### CODE REVIEW

**Comment by leplatrem** [comment]
feedback+

approach is simple and seems that it does the job!

GG :)

File: scripts/plugin.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Note: The plugin API is experimental and likely to change over time. Don't
+ * use it if you bet on stability.
+ */
+import React from "react";
+import { Route } from "react-router";
+
+
+export function flattenPluginRoutes(
```
Comment (leplatrem) [inline scripts/plugin.js]: nit: since it receives a `plugins` arguments (plural), shouldn't it be `flattenPluginsRoutes` ?

**Comment by n1k0** [comment]
I'd like a final r+ here from @leplatrem

I went with the easiest/simplest route to allow the patch to land fast regarding the way we enable plugins:

>### Enabling plugins
>
>For now and until the API stabilizes, plugins are developed in the kinto-admin repository. To enable one or more plugins, you need to set the `KINTO_ADMIN_PLUGINS` environment variable to list those you want to activate when running the development server:
>
>```
>$ KINTO_ADMIN_PLUGINS="example,signoff" npm start
>```

I've also bootstrapped the `signoff` plugin so we'll start hacking on it on this repo so we'll shape up the plugin API while discovering what's missing or doesn't work, allowing for a much faster feedback loop.

File: scripts/index.js [hunk 2]
```diff
@@ -1,5 +1,15 @@
 import renderAdmin from "./app";
-import * as testPlugin from "../sample/plugin";
 
+// Check for local plugins to enable for development.
+const plugins = (process.env.KINTO_ADMIN_PLUGINS || "")
+  .split(",")
+  .filter(x => !!x)
+  .map(pluginName => {
+    try {
+      return require(`../plugins/${pluginName}/index`);
```
Comment (n1k0) [inline scripts/index.js]: Ensure this is resolved at webpack compilation time (prod target).

**Comment by leplatrem** [comment]
r+ :)

**Pull request merged**